craft_parts/utils/partition_utils.py

```diff
@@ -51,8 +59,81 @@ def validate_partition_names(partitions: Optional[Sequence[str]]) -> None:
         raise errors.FeatureError("Partitions must be unique.")
 
+    _validate_partition_names(partitions)
+
+    _validate_namespace_conflicts(partitions)
+
+
+def _is_valid_partition(partition: str) -> bool:
+    """Check if a partition name is valid.
+
+    :param partition: partition to check
+
+    :returns: true if the namespaced partition is valid
+    """
+    return bool(re.fullmatch("[a-z]+", partition))
+
+
+def _is_valid_namespaced_partition(partition: str) -> bool:
+    """Check if a namespaced partition name is valid.
+
+    :param partition: partition to check
+
+    :returns: true if the namespaced partition is valid
+    """
+    return bool(re.fullmatch(r"([a-z]+)/(?!-)[a-z\-]+(?<!-)", partition))
+
+
+def _validate_partition_names(partitions: Sequence[str]) -> None:
+    """Validate names of partitions.
+
+    :param partitions: Sequence of partitions to validate.
+
+    :raises FeatureError: if a partition name is not valid
+    """
     for partition in partitions:
-        if not re.fullmatch("[a-z]+", partition):
+        if _is_valid_partition(partition) or _is_valid_namespaced_partition(partition):
+            continue
+
+        if "/" in partition:
             raise errors.FeatureError(
-                f"Partition {partition!r} must only contain lowercase letters."
+                message=f"Namespaced partition {partition!r} is invalid.",
+                details=(
+                    "Namespaced partitions are formatted as `<namespace>/"
+                    "<partition>`. Namespaces must only contain lowercase letters. "
+                    "Namespaced partitions must only contain lowercase letters and "
+                    "hyphens and cannot start or end with a hyphen."
+                ),
             )
+
+        raise errors.FeatureError(
+            message=f"Partition {partition!r} is invalid.",
+            details="Partitions must only contain lowercase letters.",
+        )
+
+
+def _validate_namespace_conflicts(partitions: Sequence[str]) -> None:
+    """Validate conflicts between regular partitions and namespaces.
+
+    For example, `foo` conflicts in ['default', 'foo', 'foo/bar'].
+    Assumes partition names are valid.
+
+    :raises FeatureError: for namespace conflicts
+    """
+    namespaced_partitions: Set[str] = set()
+    regular_partitions: Set[str] = set()
+
+    # sort partitions
+    for partition in partitions:
+        if _is_valid_partition(partition):
+            regular_partitions.add(partition)
+        else:
+            namespaced_partitions.add(partition)
+
+    for regular_partition in regular_partitions:
+        for namespaced_partition in namespaced_partitions:
+            if namespaced_partition.startswith(regular_partition + "/"):
+                raise errors.FeatureError(
+                    f"Partition {regular_partition!r} conflicts with the namespace of "
+                    f"partition {namespaced_partition!r}"
```

Comment:
Set operations would be nice but I see that the nested loops result in a better error message.